tests/lib/rule-tester/rule-tester.js

```diff
@@ -6183,6 +6203,518 @@ describe("RuleTester", () => {
+								},
+							],
+						},
+					);
+				});
+
+				it("should allow specifying data when there are message placeholders", () => {
+					ruleTester.run(
+						"message-id",
+						require("../../fixtures/testers/rule-tester/messageId.js")
+							.withMetaWithData,
+						{
+							assertionOptions,
+							valid: [],
+							invalid: [
+								{
+									code: "foo",
+									errors: [
+										{
+											messageId: "avoidFoo",
+											data: { name: "foo" },
+										},
+									],
+								},
+							],
+						},
+					);
+				});
+
+				it("should disallow not specifying data when there are message placeholders for errors", () => {
+					assert.throws(() => {
+						ruleTester.run(
+							"message-id",
+							require("../../fixtures/testers/rule-tester/messageId.js")
+								.withMetaWithData,
+							{
+								assertionOptions,
+								valid: [],
+								invalid: [
+									{
+										code: "foo",
+										errors: [{ messageId: "avoidFoo" }],
+									},
+								],
+							},
+						);
+					}, "Error should specify the 'data' property as the referenced message has placeholders.");
+				});
+
+				it("should allow suggestion shorthand", () => {
+					ruleTester.run(
+						"message-id",
+						require("../../fixtures/testers/rule-tester/suggestions.js")
+							.withMessageIds,
+						{
+							assertionOptions,
+							valid: [],
+							invalid: [
+								{
+									code: "foo",
+									errors: [
+										{
+											messageId: "avoidFoo",
+											data: { name: "foo" },
+											suggestions: 2,
+										},
+									],
+								},
+							],
+						},
+					);
+				});
+
+				it("should allow specifying desc for suggestions", () => {
+					ruleTester.run(
+						"message-id",
+						require("../../fixtures/testers/rule-tester/suggestions.js")
+							.withMessageIds,
+						{
+							assertionOptions,
+							valid: [],
+							invalid: [
+								{
+									code: "foo",
+									errors: [
+										{
+											message:
+												"Avoid using identifiers named 'foo'.",
+											suggestions: [
+												{
+													desc: "Rename identifier 'foo' to 'bar'",
+													output: "bar",
+												},
+												{
+													desc: "Rename identifier 'foo' to 'baz'",
+													output: "baz",
+												},
+											],
+										},
+									],
+								},
+							],
+						},
+					);
+				});
+
+				it("should allow not specifying data for referenced suggestion messages without placeholders", () => {
+					ruleTester.run(
+						"message-id",
+						require("../../fixtures/testers/rule-tester/suggestions.js")
+							.withMessageIdsWithoutPlaceholders,
+						{
+							assertionOptions,
+							valid: [],
+							invalid: [
+								{
+									code: "foo",
+									errors: [
+										{
+											message:
+												"Avoid using identifiers named 'foo'.",
+											suggestions: [
+												{
+													desc: "Rename identifier to bar",
```

Comment:
```suggestion
													messageId: "rename",
```

Shouldn't this test rather use `messageId`, as that would verify that `data` isn't required for messages that don't have placeholders regardless of `requireData`? Otherwise, all tests would pass if the `|| !hasPlaceholders` condition is removed from `assert.ok()` on line 1608 in the rule tester code.